Manage whitelisted guilds and file formats in DB. #305

atmishra opened the pull request:
This PR closes #305. This adds API to manage whitelist in the Database. Includes following changes:

1. Add Model for Whitelist
2. Add viewset, serializer for the API
3. Add migration to create existing whitelist

atmishra committed:
Add API endpoints to manage whitelist in DB

Whitelist are currently configured in default-config.yml in the bot repo.

This commit provides API to manage whitelist in the database.

atmishra committed:
Merge branch 'master' of https://github.com/python-discord/site into whitelist-api

atmishra committed:
Add data migration to add existing whitelist to db

atmishra committed:
Make migration reversible

atmishra committed:
Fix delete query

atmishra committed:
Add Whitelist model to model test cases

jchristgit commented on pydis_site/apps/api/urls.py:
```diff
@@ -7,7 +7,7 @@
     LogEntryViewSet, NominationViewSet,
     OffTopicChannelNameViewSet, ReminderViewSet,
-    RoleViewSet, TagViewSet, UserViewSet
+    RoleViewSet, TagViewSet, UserViewSet, WhitelistViewSet
```
the other lines are all 2 items per line, can you move `UserViewSet, WhitelistViewSet` on a new line?

jchristgit commented on pydis_site/apps/api/models/bot/whitelist.py:
```diff
@@ -0,0 +1,35 @@
+    INVITE = "invite"
+    EXTENSION = "extension"
+    CHANNEL = "channel"
+    ROLE = "role"
```
hmm, what's the reason for these constants? I don't see them used anywhere else, but if they should be, I think a separate Enum would be a better fit (e.g. `WhitelistType`)

atmishra replied:
Removed the constants.

jchristgit replied:
Hmm, did you forget to push, or am I looking at old code here? Constants are still here.

atmishra replied:
My Bad! I removed the use of the constants but forgot to remove them.

jchristgit reviewed (changes requested):
Sorry for the late review. Can you

- Resolve the merge conflicts
- Rename the  `0050_auto_20191117_1821.py` migration to something more meaningful, e.g. `0050_import_whitelisted_items`

the icing on the cake would be if the data import migration would only delete the imported items, but I don't think that's necessary

atmishra committed:
Merge branch 'master' into whitelist-api

atmishra committed:
Resolve merge conflict & add PR review changes

Changed default migration file name to meaningful name, removed unused constants

atmishra committed:
Add unique together constraint for type and whitelist_item

This will prevent duplicate items in a single type

atmishra committed:
Remove unused constants

atmishra committed:
Merge branch 'master' of https://github.com/python-discord/site into whitelist-api

atmishra committed:
Merge confilicting migration

jchristgit reviewed (approved):
Code-wise this looks great!

Shortly before a merge, we need to update the whitelist to match what's currently in the bot's configuration file, since the data migration and the configuration file have drifted apart a bit since the PR was opened. I think that should wait until the second review though (or with a second data migration after merge).

MarkKoz commented on pydis_site/apps/api/migrations/0053_merge_20200114_1559.py:
```diff
@@ -0,0 +1,14 @@
+        ('api', '0052_add_whitelist_unique_constraint'),
+    ]
+
+    operations = [
+    ]
```
What is the purpose of this migration if it has no operations?

lemonsaurus replied:
To merge two heads, and restore linear order to the migration history.

MarkKoz commented on pydis_site/apps/api/models/bot/whitelist.py:
```diff
@@ -0,0 +1,36 @@
+    """Whitelisted item for a type."""
+
+    TYPE_CHOICES = (
+        ("invite", "invite"),
+        ("extension", "extension"),
+        ("channel", "channel"),
+        ("role", "role")
```
Capitalise the second element of each tuple - those are the human-readable names.
```suggestion
        ("invite", "Invite"),
        ("extension", "Extension"),
        ("channel", "Channel"),
        ("role", "Role")
```
Not sure if this requires a migration.

MarkKoz commented on pydis_site/apps/api/models/bot/whitelist.py:
```diff
@@ -0,0 +1,36 @@
+    type = models.CharField(
+        choices=TYPE_CHOICES,
+        max_length=100,
+        help_text=(
+            "Type of the whitelisted item."
+        ),
```
```suggestion
        help_text="Type of the whitelisted item.",
```

MarkKoz commented on pydis_site/apps/api/models/bot/whitelist.py:
```diff
@@ -0,0 +1,36 @@
+    )
+    whitelisted_item = models.CharField(
+        max_length=300,
+        help_text=(
+            "whitelisted item"
+        )
```
```suggestion
        help_text="Whitelisted item."
```

MarkKoz commented on pydis_site/apps/api/models/bot/whitelist.py:
```diff
@@ -0,0 +1,36 @@
+            "Type of the whitelisted item."
+        ),
+    )
+    whitelisted_item = models.CharField(
+        max_length=300,
```
This should be a `TextField` without any sort of length restriction.

MarkKoz commented on pydis_site/apps/api/models/bot/whitelist.py:
```diff
@@ -0,0 +1,36 @@
+        ("channel", "channel"),
+        ("role", "role")
+    )
+    type = models.CharField(
+        choices=TYPE_CHOICES,
+        max_length=100,
```
This should be a `TextField` without any sort of length restriction.

MarkKoz commented on pydis_site/apps/api/viewsets/bot/whitelist.py:
```diff
@@ -0,0 +1,69 @@
+    View providing create, list and  delete operations on whitelist shown by our bot.
+
+    ## Routes
+    ### GET /bot/whitelist?type=invites
```
Don't include a query parameter as part of the route. Document the query parameters in a separate section.

MarkKoz commented on pydis_site/apps/api/viewsets/bot/whitelist.py:
```diff
@@ -0,0 +1,69 @@
+    GenericViewSet,
+):
+    """
+    View providing create, list and  delete operations on whitelist shown by our bot.
```
This suggestion assumes you've added `RetrieveModelMixin` and thus made it support CRUD.

```suggestion
    View providing CRUD operations on the whitelist used by the Discord bot's filters.
```

MarkKoz commented on pydis_site/apps/api/viewsets/bot/whitelist.py:
```diff
@@ -0,0 +1,69 @@
+
+
+class WhitelistViewSet(
+    mixins.CreateModelMixin,
+    mixins.ListModelMixin,
+    mixins.DestroyModelMixin,
+    GenericViewSet,
```
Wouldn't hurt to add `RetrieveModelMixin` and just make this CRUD. May come in handy in the future and it's a 1 line code change (well, besides the docstring) thanks to DRF's magic.

MarkKoz commented on pydis_site/apps/api/viewsets/bot/whitelist.py:
```diff
@@ -0,0 +1,69 @@
+    ...     'type': str,
+    ...     'whitelisted_item': str
+    ... }
+
```
There was an extra blank line.
```suggestion
```

MarkKoz commented on pydis_site/apps/api/models/bot/whitelist.py:
```diff
@@ -0,0 +1,36 @@
+
+
+class Whitelist(ModelReprMixin, models.Model):
+    """Whitelisted item for a type."""
```
```suggestion
    """An item of a certain type which is excluded from the Discord bot's filters."""
```

MarkKoz commented on pydis_site/apps/api/migrations/0051_import_whitelisted_items.py:
```diff
@@ -0,0 +1,77 @@
+class Migration(migrations.Migration):
+    def create_whitelist(app, schema_editor):
+        Whitelist = app.get_model("api", "Whitelist")
+        invites_whitelist = [
```
Let's update this migration with the latest configuration from the bot before this PR is merged.

MarkKoz commented on pydis_site/apps/api/models/bot/whitelist.py:
```diff
@@ -0,0 +1,36 @@
+
+    TYPE_CHOICES = (
+        ("invite", "invite"),
+        ("extension", "extension"),
```
This should be renamed to more clearly express that it's a file extension and not, say, a Discord extension. My suggestion: `("file_ext", "File Extension"),`

lemonsaurus reviewed (commented):
This has been open without any changes for long enough.
